frontend/src/components/widgets/DataFrame/DataFrame.tsx

```diff
@@ -0,0 +1,375 @@
+        } as GridCell
+      }
+
+      if (col > columns.length - 1) {
+        // This should never happen
+        return getCellTemplate(ColumnType.Text, true)
+      }
+
+      const cellTemplate = columns[col].getTemplate()
+      if (row > numRows - 1) {
+        // This should never happen
+        return cellTemplate
+      }
+      try {
+        // Quiver has the header in first row
+        const quiverCell = element.getCell(row + 1, col)
+        return fillCellTemplate(cellTemplate, quiverCell, element.cssStyles)
+      } catch (error) {
+        // This should not happen in read-only table.
+        logError(error)
+        return cellTemplate
+      }
+    },
+    [columns, numRows, element]
+  )
+
+  const { getCellContent: getCellContentSorted } = useColumnSort({
+    columns,
+    getCellContent,
+    rows: numRows,
+    sort,
+  })
+
+  if (sort !== undefined) {
+    columns = columns.map(column => {
+      if (column.id === sort.column.id) {
+        return {
+          ...column,
+          title:
+            sort.direction === "asc"
+              ? `↑ ${column.title}`
+              : `↓ ${column.title}`,
+        }
+      }
+      return column
+    })
+  }
+
+  return {
+    numRows,
+    numIndices,
+    columns,
+    getCellContent: getCellContentSorted,
+    onColumnResize,
+  }
+}
+export interface DataFrameProps {
+  element: Quiver
+  height?: number
+  width: number
+}
+
+function DataFrame({
+  element,
+  height: propHeight,
+  width: propWidth,
+}: DataFrameProps): ReactElement {
+  const [sort, setSort] = React.useState<ColumnSortConfig>()
+
+  const {
+    numRows,
+    numIndices,
+    columns,
+    getCellContent,
+    onColumnResize,
+  } = useDataLoader(element, sort)
+
+  const [isFocused, setIsFocused] = React.useState<boolean>(true)
+
+  const [gridSelection, setGridSelection] = React.useState<GridSelection>({
+    columns: CompactSelection.empty(),
+    rows: CompactSelection.empty(),
+  })
+
+  const dataEditorRef = React.useRef<DataEditorRef>(null)
+
+  const minWidth = MIN_COLUMN_WIDTH + 3
+
+  const onHeaderClick = React.useCallback(
+    (index: number) => {
+      let sortDirection = "asc"
+      const clickedColumn = columns[index]
+
+      if (sort && sort.column.id === clickedColumn.id) {
+        // The clicked column is already sorted
+        if (sort.direction === "asc") {
+          // Sort column descending
+          sortDirection = "desc"
+        } else {
+          // Remove sorting of column
+          setSort(undefined)
+          return
+        }
+      }
+
+      setSort({
+        column: clickedColumn,
+        direction: sortDirection,
+        mode: getColumnSortMode(
+          (clickedColumn as GridColumnWithCellTemplate).columnType
+        ),
+      } as ColumnSortConfig)
+    },
+    [sort, columns]
+  )
+
+  // Calculate min height for the resizable container. header + one column, and +3 pixels for borders
+  const minHeight = 2 * ROW_HEIGHT + 3
```

Comment:
I added a comment to make the +3 a bit more clear:

```
  // Calculate min width for the resizable table container.
  // Based on one column at minimum width + 2 for borders + 1 to prevent overlap problem with selection ring.
  const minWidth = MIN_COLUMN_WIDTH + 3

  // Calculate min height for the resizable table container.
  // Based on header + one column, and + 2 for borders + 1 to prevent overlap problem with selection ring.
  const minHeight = 2 * ROW_HEIGHT + 3
```

But I will think about moving that to constants. In a previous version, I had a bit more dynamic with calculating `minWidth` and `minHeight`. But that is not in there anymore.